sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala

```diff
@@ -17,285 +17,159 @@
-  override def dataType: DataType = base.dataType
+  // Specify the BoundReference for Aggregate Buffer
+  def initialBoundReference(buffers: Seq[BoundReference]): Unit
 
-  def update(input: Row): Unit
+  // Initialize (reinitialize) the aggregation buffer
+  def reset(buf: MutableRow): Unit
 
-  // Do we really need this?
-  override def newInstance(): AggregateFunction = {
-    makeCopy(productIterator.map { case a: AnyRef => a }.toArray)
-  }
-}
+  // Expect the aggregate function fills the aggregation buffer when
+  // fed with each value in the group
+  def iterate(arguments: Any, buf: MutableRow): Unit
 
-case class Min(child: Expression) extends PartialAggregate with trees.UnaryNode[Expression] {
+  // Merge 2 aggregation buffer, and write back to the later one
+  def merge(value: Row, buf: MutableRow): Unit
 
-  override def nullable: Boolean = true
-  override def dataType: DataType = child.dataType
-  override def toString: String = s"MIN($child)"
-
-  override def asPartial: SplitEvaluation = {
-    val partialMin = Alias(Min(child), "PartialMin")()
-    SplitEvaluation(Min(partialMin.toAttribute), partialMin :: Nil)
-  }
-
-  override def newInstance(): MinFunction = new MinFunction(child, this)
-}
-
-case class MinFunction(expr: Expression, base: AggregateExpression) extends AggregateFunction {
-  def this() = this(null, null) // Required for serialization.
-
-  val currentMin: MutableLiteral = MutableLiteral(null, expr.dataType)
-  val cmp = GreaterThan(currentMin, expr)
-
-  override def update(input: Row): Unit = {
-    if (currentMin.value == null) {
-      currentMin.value = expr.eval(input)
-    } else if(cmp.eval(input) == true) {
-      currentMin.value = expr.eval(input)
-    }
-  }
-
-  override def eval(input: Row): Any = currentMin.value
-}
-
-case class Max(child: Expression) extends PartialAggregate with trees.UnaryNode[Expression] {
-
-  override def nullable: Boolean = true
-  override def dataType: DataType = child.dataType
-  override def toString: String = s"MAX($child)"
-
-  override def asPartial: SplitEvaluation = {
-    val partialMax = Alias(Max(child), "PartialMax")()
-    SplitEvaluation(Max(partialMax.toAttribute), partialMax :: Nil)
-  }
-
-  override def newInstance(): MaxFunction = new MaxFunction(child, this)
-}
-
-case class MaxFunction(expr: Expression, base: AggregateExpression) extends AggregateFunction {
-  def this() = this(null, null) // Required for serialization.
-
-  val currentMax: MutableLiteral = MutableLiteral(null, expr.dataType)
-  val cmp = LessThan(currentMax, expr)
-
-  override def update(input: Row): Unit = {
-    if (currentMax.value == null) {
-      currentMax.value = expr.eval(input)
-    } else if(cmp.eval(input) == true) {
-      currentMax.value = expr.eval(input)
-    }
-  }
+  // Semantically we probably don't need this, however, we need it when
+  // integrating with Hive UDAF(GenericUDAF)
+  @deprecated
+  def terminatePartial(buf: MutableRow): Unit = {}
```

Comment:
We can not do that, as they are quite different semantics in Hive, being compatible with Hive behavior is quite critical mission for this PR.